File: outco/stringAndArray/pigLatin.js
// const { PerformanceObserver, performance } = require("perf_hooks");
var t0 = performance.now();

// Pig latin translation uses the following rules:
// - for words that start with a vowel, add 'yay' to the end
// - for words that start with a nonvowel,
// move all letters before the first vowel to the end of the word and add 'ay'

// input: a word
// output: a word 
// constraints: O(N) run time, O(N) space complexity
// edge cases: return "" for empty strings

// Succesfully tested the solution on 12/10/22.
const pigLatin = (word) => {
  if (word.length === 0) {
    return ""
  }
  let arr = word.split('')
  // if the array at index 0 is a vowel
  if (arr[0] == "a" || arr[0] =="e"|| arr[0] == "i" || arr[0] == "o" || arr[0] == "u"){
    return word + "yay"
  }
  // loop until a vowel is found at index i
  for (let i = 0; i < arr.length; i++) {
        if (arr[i] == "a" || arr[i] =="e"|| arr[i] == "i" || arr[i] == "o" || arr[i] == "u") {
          let fragment = arr.slice(0, i);
          let main = arr.slice(i, word.length+1);
          return main.join("") + fragment.join("") + "ay";
        }
      }
};

console.log(pigLatin("apple")); // => "appleyay"
console.log(pigLatin("")); // => ""
console.log(pigLatin("eat")); // => "eatyay"
console.log(pigLatin("banana")); // => "ananabay"
console.log(pigLatin("trash")); // => "ashtray"

var t1 = performance.now();
console.log("Call to pigLatin took " + (t1 - t0) + " milliseconds.");

File: c++journey/dsaDomains/stringAndArray/pigLatin.cpp
/* 
* Pig latin translation uses the following rules:
* - for words that start with a vowel, add 'yay' to the end
* - for words that start with a nonvowel,
* move all letters before the first vowel to the end of the word and add 'ay'

* input: a word
* output: a word 
* constraints: O(N) run time, O(N) space complexity
* edge cases: return "" for empty strings

* 
*/
#include <vector> 
#include <iostream>
using namespace std;

string pigLatin(string inputStr){
    string vowelsStr = "aeiouy";
    // words which start with vowel
    for (char &charG: vowelsStr){
        // cout << "inputStr..." << inputStr << endl;
        // cout << "charG.." << charG << endl;
        // cout << "inputStr[0].." << inputStr[0] << endl;
        if (charG == inputStr[0]){
            // add 'yay' to the end
            return inputStr+"yay";
        }
    }
    // words which start with consonant
    int intG = 0;
    while (intG < inputStr.size()){
        for (char charG: vowelsStr){
            // found the first vowel
            if (inputStr[intG] == charG){
                // chop the string from the start until the letter before the first
                // consonant
                string fragmentStr = inputStr.substr(0, intG);
                // chop the string from the vowel to the end of the string
                string mainStr = inputStr.substr(intG);
                // append the fragment to the main string with "ay"
                return mainStr + fragmentStr + "ay";
            }
        }
        intG++;
    }
    return "";
}

int main(){
    string inputStr1 = "apple";
    string inputStr2 = "";
    string inputStr3 = "eat";
    string inputStr4 = "banana";
    string inputStr5 = "trash";
    string inputStr6 = "yellow";
    string inputStr7 = "why";

    string outputStr1 = pigLatin(inputStr1);
    string outputStr2 = pigLatin(inputStr2);
    string outputStr3 = pigLatin(inputStr3);
    string outputStr4 = pigLatin(inputStr4);
    string outputStr5 = pigLatin(inputStr5);
    string outputStr6 = pigLatin(inputStr6);
    string outputStr7 = pigLatin(inputStr7);

    cout << "Output str 1: " << outputStr1 << endl;
    cout << "Output str 2: " << outputStr2 << endl;
    cout << "Output str 3: " << outputStr3 << endl;
    cout << "Output str 4: " << outputStr4 << endl;
    cout << "Output str 5: " << outputStr5 << endl;
    cout << "Output str 6: " << outputStr6 << endl;
    cout << "Output str 7: " << outputStr7 << endl;
    return 0;
}